[Auto Downloading] Update UI for Podcast Auto Downloading

mebarbosa opened the pull request:
## Description
- This adds the auto download settings to podcast setting
- Persist the download limit in database d0b94ac7cd763b924628274e217ce209d44221ca
- Figma: [VvERCFIopLjFL8Fnczsf8r-fi-1_756](https://href.li/?https://www.figma.com/file/VvERCFIopLjFL8Fnczsf8r/?node-id=1-756)



**Note:** Default behavior: Two latest episodes
**Note2**: I will align and address in another PR with designs if we should reset the auto download limit to two latest download limit after toggle `Auto Download` off
**Note 3**: I will address in another PR the database update after the user subscribes to a podcast

Fixes #2971 

## Testing Instructions

### AUTO_DOWNLOAD feature flag OFF
1. Have the FF disabled
2. Go to some podcast -> Subscribe to this podcast -> Tap on Settings
3. Toggle the `Auto Download` ON
4. ✅ Ensure you don't see the limit download section

### AUTO_DOWNLOAD feature flag ON
1. Have the FF enabled
2. Go to some podcast -> Subscribe to this podcast -> Tap on Settings
3. Toggle the `Auto Download` ON
4. ✅ Ensure you see the limit download section
5. Tap on limit downloads label and select one option
6. ✅ Ensure your option is selected
8. ✅ Ensure 🔵 Tracked: podcast_settings_auto_download_limit_toggled, Properties: {"value":... is tracked

## Screenshots or Screencast 

<img src="https://github.com/user-attachments/assets/dbe249ab-59c9-4bf7-8dee-260c51b98718" width="300">


## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [x] with different themes
- [x] with a landscape orientation
- [x] with the device set to have a large display and font size
- [x] for accessibility with TalkBack

dangermattic commented:
<!--
  1 Error: Please add tests for class <co...
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>PodcastAutoDownloadLimitConverter</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

CookieyedCodes commented:
Just checking with this auto download limit will it be completely customisable?, be able to be based on the sort order? and will it not delete downloads/archive episodes from feeds that are already have more episides downloaded?

MiSikora commented on modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/db/AppDatabase.kt:
```diff
@@ -841,6 +843,10 @@ abstract class AppDatabase : RoomDatabase() {
             }
         }
 
+        val MIGRATION_100_101 = addMigration(100, 101) { database ->
+            database.execSQL("ALTER TABLE episode ADD COLUMN auto_download_limit INTEGER DEFAULT 2")
+        }
```
This migration is incorrect.

1. It doesn't update `auto_download_limit` for existing podcasts.
2. The definition doesn't match `Podcast` entity. The entity doesn't define default value on the SQL level but on the Kotlin level so there no `DEFAULT 2` when the table is created without migration.
3. The migration crashes the app due to a typo. It should be `TABLE podcasts` .instead of `TABLE episode`.

MiSikora commented on modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/entity/Podcast.kt:
```diff
@@ -59,6 +60,7 @@ data class Podcast(
     @ColumnInfo(name = "show_notifications") var isShowNotifications: Boolean = false,
     @ColumnInfo(name = "show_notifications_modified") var showNotificationsModified: Date? = null,
     @ColumnInfo(name = "auto_download_status") var autoDownloadStatus: Int = 0,
+    @ColumnInfo(name = "auto_download_limit") var autoDownloadLimit: AutoDownloadLimitSetting = AutoDownloadLimitSetting.TWO_LATEST_EPISODE,
```
This default value leads to some significant issues. When I sign in and to a clean sync all of the podcasts have `auto_download_limit` set to `TWO_LATEST_EPISODES`. Moreover, all podcasts that I see in discover have this column set to `TWO_LATEST_EPISODES`.

mebarbosa replied:
I am going to update it to default OFF and change to TWO_LATEST_EPISODE after the user subscribes to the podcast

MiSikora commented on modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/type/AutoDownloadLimitSetting.kt:
```diff
@@ -50,10 +50,12 @@ enum class AutoDownloadLimitSetting(
             return try {
                 val intValue = stringValue.toInt()
-                entries.firstOrNull { it.preferenceInt == intValue } ?: TWO_LATEST_EPISODE
+                entries.firstOrNull { it.id == intValue } ?: TWO_LATEST_EPISODE
```
I think it would be better to change this method to return `AutoDownloadLimitSetting?` instead. The current default value, `TWO_LATEST_EPISODES`, mixes different use cases and could lead to potential errors when we need different behaviors.

I suggest removing the default value and determining the appropriate setting at the call sites instead.

MiSikora commented on modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/type/AutoDownloadLimitSetting.kt:
```diff
@@ -50,10 +50,12 @@ enum class AutoDownloadLimitSetting(
             }
         }
+
+        fun fromInt(id: Int) = (AutoDownloadLimitSetting.entries.firstOrNull { it.id == id } ?: TWO_LATEST_EPISODE)
```
Similarly to `fromPreferenceString` I think I'd be better to use any default value in case of missing entries.

mebarbosa commented:
@MiSikora I addressed all changes. 

After implementing them I tried to run the migration and I had to do it twice. When I checked to `main` and build the app, I realized the code from the previous branch was installed. I cleaned build, deleted manually the local cache and the old code was still there. So I run again with `./gradlew :app:installDebug --rerun-tasks` command and finally was able to get the `main` code installed

MiSikora commented on modules/features/podcasts/src/main/java/au/com/shiftyjelly/pocketcasts/podcasts/view/podcast/PodcastSettingsFragment.kt:
```diff
@@ -490,6 +517,18 @@ class PodcastSettingsFragment : BasePreferenceFragment(), FilterSelectFragment.L
 
+    private fun updateAutoDownloadLimit(podcast: Podcast) {
+        val options = AutoDownloadLimitSetting.entries
+        val podcastAutoDownloadLimit = podcast.autoDownloadLimit ?: AutoDownloadLimitSetting.TWO_LATEST_EPISODE
```
Nothing to change here right now but it will be important to remember to use `TWO_LATEST_EPISODE` as the default value when doing actual downloading.

MiSikora reviewed (approved):
I think there is a UX issue. It doesn't make sense to me to have an `Off` setting when there is a toggle to disable auto downloads. What does it mean to enable the setting and have the limit set to `Off`?

![share_8631750866837509173](https://github.com/user-attachments/assets/4a5fa42f-5643-490d-beaf-a19602b39165)

mebarbosa commented:
> I think there is a UX issue. It doesn't make sense to me to have an `Off` setting when there is a toggle to disable auto downloads. What does it mean to enable the setting and have the limit set to `Off`? If it means that there is no limit there is a separate value for it.
> 
> Off	All
> ![share_8631750866837509173](https://github.com/user-attachments/assets/4a5fa42f-5643-490d-beaf-a19602b39165)	![share_2201720753321164508](https://github.com/user-attachments/assets/fb8acb88-ea2d-4603-b808-75e44807042a)

@MiSikora I agree. It does not make sense to me to have this toggle. I reported this here p1728473112997409/1728311298.538509-slack-C07QMC2GB1B and I am waiting for the response. In the next PR I can address it

mebarbosa commented:
> Just checking with this auto download limit will it be completely customisable?, be able to be based on the sort order? and will it not delete downloads/archive episodes from feeds that are already have more episides downloaded?

@CookieyedCodes We want to implement something to download X latest episodes for now. 
For the deletion we are building some rules for this to exclude some cases

CookieyedCodes commented:
Hmmmmm while I could see that as useful I think a per podcast auto download that is tied to the list sort order & ignoreing archived episodes would be more helpful imo because the vast majority of podcasts are listened to in backlog imo & once caught up on it would behave like the current auto download functionality and as the default is newest to oldest it would automatically have that functionality, even a toggle between both options latest and list would be a massive improvement 🤔
